File: back-end/TaskManagement/BLL/PresenceHoursService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using BOL;
using MySql.Data.MySqlClient;
using DAL;

namespace BLL
{
    public static class PresenceHoursService
    {
        public static List<PresenceHours> GetPresenceHours(int workerId, int projectId)
        {
            try
            {
                string query = $"SELECT p.*,u.user_name FROM task_management.presence_hours p JOIN task_management.user u ON p.worker_id=u.user_id WHERE end_hour IS NOT NULL AND worker_id={workerId} AND project_id={projectId};";

                Func<MySqlDataReader, List<PresenceHours>> InitPresenceHoursList = (reader) =>
                {
                    List<PresenceHours> presencesHours = new List<PresenceHours>();
                    while (reader.Read())
                    {
                        PresenceHours presenceHours = BaseService.initPresenceHours(reader);
                        presencesHours.Add(presenceHours);
                    }
                    return presencesHours;
                };

                List<PresenceHours> presenceHoursList = DBAccess.RunReader(query, InitPresenceHoursList);
                return presenceHoursList.Where(presenceHours => presenceHours.StartHour.Month == DateTime.Today.Month).ToList();
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public static int GetPresenceHoursSum(int projectId, int workerId)
        {
            string query = "SELECT SUM(TIMESTAMPDIFF(HOUR, start_hour, end_hour)) " +
                           "FROM task_management.presence_hours " +
                           $"WHERE project_id = {projectId} AND worker_id = {workerId}";
            int sum = (int)DBAccess.RunScalar(query);
            return sum;
        }
        //public static List<PresenceHours> GetPresenceHoursByProjectId(int projectId)
        //{
        //    try
        //    {
        //        List<PresenceHours> presenceHoursList = GetPresenceHours().Where(presenceHour => presenceHour.ProjectId == projectId).ToList();
        //        return presenceHoursList;
        //    }
        //    catch (Exception ex)
        //    {
        //        throw ex;
        //    }
        //}


        public static int AddPresenceHours(PresenceHours newPresenceHours)
        {
            try
            {
                newPresenceHours.StartHour = newPresenceHours.StartHour.ToLocalTime();
                string query = $"INSERT INTO task_management.presence_hours(worker_id,project_id,start_hour) VALUES({newPresenceHours.WorkerId}, {newPresenceHours.ProjectId}, {BaseService.FormatDate(newPresenceHours.StartHour, "yyyy-MM-dd HH:mm:ss")});";
                if (DBAccess.RunNonQuery(query) == 1)
                {
                    query = $"SELECT * FROM task_management.presence_hours ORDER BY presence_hours_id  DESC LIMIT 1;";
                    int id = Convert.ToInt32(DBAccess.RunScalar(query));
                    return id;
                }
                return -1;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public static bool EditPresenceHours(PresenceHours presenceHours)
        {
            try
            {
                string query = $"UPDATE task_management.presence_hours SET end_hour={BaseService.FormatDate(((DateTime)presenceHours.EndHour).ToLocalTime(), "yyyy-MM-dd HH:mm:ss")} WHERE presence_hours_id={presenceHours.PresenceHoursId};";
                bool created = DBAccess.RunNonQuery(query) == 1;
                if (created)
                {
                    WorkerHours workerHours = WorkerHoursService.GetWorkerHoursPerProject(presenceHours.WorkerId, presenceHours.ProjectId)[0];
                    int presenceHoursSum = GetPresenceHoursSum(presenceHours.WorkerId, presenceHours.ProjectId);
                    if (presenceHoursSum >= workerHours.NumHours)
                    {
                        workerHours.IsComplete = true;
                        WorkerHoursService.EditWorkerHours(workerHours);
                        ProjectService.updateProjectStatus(presenceHours.ProjectId);
                    }

                }
                return true;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public static List<dynamic> GetPresenceStatusPerWorkers(int teamLeaderId)
        {
            string query =
            "SELECT user_name, SUM(TIMESTAMPDIFF(HOUR, start_hour, end_hour)) " +
            "FROM task_management.user u " +
            "JOIN task_management.presence_hours p " +
            "ON u.user_id = p.worker_id " +
            $"WHERE team_leader_id={teamLeaderId} " +
            "GROUP BY user_id;";
            Func<MySqlDataReader, List<dynamic>> InitPresenceStatusList = (reader) =>
            {
                List<dynamic> presenceStatusList = new List<dynamic>();
                while (reader.Read())
                {
                    var presenceStatus = new { userName = reader.GetString(0), presenceHours = reader.GetInt32(1) };
                    presenceStatusList.Add(presenceStatus);
                }
                return presenceStatusList;
            };
            return DBAccess.RunReader(query, InitPresenceStatusList);
        }
        public static List<dynamic> GetPresenceStatusPerProjects(int workerId)
        {
            //todo
            string query =
                //pre.project_id
                "SELECT pro.project_name,SUM(TIMESTAMPDIFF(HOUR, start_hour, end_hour)),num_hours " +
                "FROM task_management.presence_hours pre " +
                "JOIN task_management.project pro " +
                "ON pro.project_id = pre.project_id " +
                "JOIN task_management.worker_hours w " +
                "ON pro.project_id = w.project_id " +
                $"WHERE pre.worker_id = {workerId} AND w.worker_id = {workerId} " +
                "GROUP BY pre.project_id;";
            Func<MySqlDataReader, List<dynamic>> InitPresenceStatusList = (reader) =>
            {
                List<dynamic> presenceStatusList = new List<dynamic>();
                while (reader.Read())
                {
                    var presenceStatus = new { projectName = reader.GetString(0), presenceHours = reader.GetInt32(1), projectHours = reader.GetInt32(2) };
                    presenceStatusList.Add(presenceStatus);
                }
                return presenceStatusList;
            };
            return DBAccess.RunReader(query, InitPresenceStatusList);
        }
    }
}

File: back-end/php/task_management/bll/presence_hours_service.php

<?php

class presence_hours_service extends base_service {

    public function add_presence_hours($new_presence_hours) {
//check if date and time is exactly
        $query = "INSERT INTO task_management.presence_hours(worker_id,project_id,start_hour) " .
                "VALUES('{$new_presence_hours['workerId']}', '{$new_presence_hours['projectId']}', " .
                "{$this->format_date($new_presence_hours['startHour'], 'Y-m-d H:i:s')});";

        $presence_hours_id = db_access::run_non_query($query);
        return isset($presence_hours_id) ? $presence_hours_id : -1;
    }

    public function edit_presence_hours($presence_hours) {
        //check if date and time is exactly
        $query = "UPDATE task_management.presence_hours " .
                "SET end_hour={$this->format_date($presence_hours['endHour'], 'Y-m-d H:i:s')} " .
                "WHERE presence_hours_id={$presence_hours['presenceHoursId']};";

        $created = db_access::run_non_query($query) == 1;
        if ($created) {
            $worker_hours_service = worker_hours_service::get_instance();
            $worker_hours = $worker_hours_service->get_worker_hours_per_project($presence_hours['workerId'], $presence_hours['projectId'])[0];
            $worker_hours_service->edit_worker_hours($worker_hours);
        }
        return $created;
    }

    public function get_presence_status_per_workers($team_leader_id) {
        //create view that select presence status
        $query = "CREATE VIEW task_management.presence_status " .
                "AS " .
                "SELECT user_name, pro.project_name,w.num_hours, IFNULL(SUM(TIMESTAMPDIFF(SECOND, start_hour, end_hour) / 3600), 0) AS presence " .
                "FROM task_management.user u " .
                "JOIN task_management.worker_hours w ON w.worker_id = u.user_id " .
                "JOIN task_management.project pro ON w.project_id = pro.project_id " .
                "LEFT JOIN task_management.presence_hours pre ON pre.project_id = pro.project_id AND pre.project_id = w.project_id AND pre.worker_id = u.user_id AND pre.worker_id = w.worker_id " .
                "WHERE u.team_leader_id = $team_leader_id AND(MONTH(start_hour) is null or MONTH(start_hour) = MONTH(CURRENT_DATE())) GROUP BY user_name, pre.project_id;";
        db_access::run_non_query($query);
//select sum of all projects presence status per worker and details of every project
        $query = "SELECT user_name,null AS project_name,sum(num_hours) AS num_hours, sum(presence)  AS presence " .
                "FROM task_management.presence_status " .
                "GROUP BY user_name " .
                "UNION SELECT * FROM task_management.presence_status;";
        $presence_status_list = db_access:: run_reader($query, function ($model) {
                    return $this->init_presence_status($model);
                });
        $query = "DROP VIEW task_management.presence_status";
        db_access::run_non_query($query);

        return $presence_status_list;
    }

    public function get_presence_status_per_projects($worker_id) {
        $query = "SELECT pro.project_name,IFNULL(SUM(TIMESTAMPDIFF(SECOND, start_hour, end_hour) / 3600), 0) AS presence, num_hours " .
                "FROM task_management.worker_hours w " .
                "LEFT JOIN task_management.project pro " .
                "ON pro.project_id = w.project_id " .
                "LEFT JOIN task_management.presence_hours pre " .
                "ON pro.project_id = pre.project_id AND pre.worker_id = $worker_id AND MONTH(start_hour) = MONTH(CURRENT_DATE()) " .
                "WHERE w.worker_id = $worker_id " .
                "AND w.num_hours > 0 " .
                "GROUP BY pre.project_id;";
        $presence_status_list = db_access::run_reader($query, function($model) {
                    return $this->init_presence_status($model);
                });
        return $presence_status_list;
    }

    public function get_presence_hours_sum($project_id, $worker_id) {
        $query = "SELECT IFNULL(SUM(TIMESTAMPDIFF(SECOND, start_hour, end_hour)/3600),0) AS sum " .
                "FROM task_management.presence_hours " .
                "WHERE project_id = $project_id AND worker_id = $worker_id";
        $sum = db_access::run_scalar($query);
        return $sum;
    }

    public function get_presence_hours($worker_id, $project_id) {
        $query = "SELECT p.*,u.user_name FROM task_management.presence_hours p " .
                "JOIN task_management.user u ON p.worker_id=u.user_id " .
                "WHERE end_hour IS NOT NULL AND worker_id=$worker_id AND project_id=$project_id;";
        $presence_hours_list = db_access:: run_reader($query, function ($model) {
                    return $this->init_presence_hours($model);
                });
        return $presence_hours_list;
    }

    private function init_presence_status($presence_status) {
        $new_presence_status = array();
        if (array_key_exists('user_name', $presence_status)) {
            $new_presence_status['userName'] = $presence_status['user_name'];
        }
        $new_presence_status['projectName'] = $presence_status['project_name'];
        $new_presence_status['projectHours'] = $presence_status['num_hours'];
        $new_presence_status['presenceHours'] = $presence_status['presence'];
        return $new_presence_status;
    }

}
